### PR TITLE
Added ability to search by static image URL

### PR DESCRIPTION
**Pull request by gabeklavans**
I added in the ability to send the bot a URL to a static image and have it return results from trace.moe using the URL parameter in the official API. The bot will parse a URL out of any message, but will only send the first URL it finds to the trace.moe API. If both an image file and URL are sent in the same message, it will prioritize the image file.
I wanted to do as little refactoring as possible, so there may be some duplicate code that can be turned into functions, if desired.

Tests I did:
- Send file to bot through private and group message (to make sure it still worked)
- Send URL to bot through PM and through group message with @ mention
- Forward image to bot through PM
- Forward URL to bot through PM
- Forward image or URL to group message with @ mention in the message already

I tested using polling instead of webhooks (because I'm not really sure how webhooks work yet haha), but I don't think this should affect the functionality. I also tested without an API token from trace.moe, but it seemed to work fine without one. I left the code in for injecting the token into the URL request, still.

I also updated the bot messages and such to reflect the changes :D

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -69,11 +82,14 @@ const submitSearch = (buffer) =>
       return;
     }
     if (!searchResult.docs) {
-      resolve({ text: "Backend server error, please try again later." });
+      resolve({
+        text:
+          "Error with search response. Is your query correctly formatted? Note: URL searches must be for static images, not gifs",
```
Comment (soruly) [inline server.js:87]: I know the original error message is a bit vague. But this is even more confusing if it shows messages about image URLs to all users (even they they are not searching by URL).

File: server.js [hunk 2]
```diff
@@ -38,24 +42,33 @@ const formatTime = (timeInSeconds) => {
...
-    form.append("image", buffer, "blob");
     let response = {};
-    try {
-      response = await fetch(`https://trace.moe/api/search?token=${TRACE_MOE_TOKEN}`, {
-        body: form,
-        method: "POST",
-      });
-    } catch (error) {
-      reject(error);
+    if (kind === "url") {
```
Comment (soruly) [inline server.js:52]: What about using [Buffer.isBuffer()](https://nodejs.org/api/buffer.html#buffer_class_method_buffer_isbuffer_obj) to determine which type of search it's using?

File: server.js [hunk 3]
```diff
@@ -189,11 +205,48 @@ const limitExceeded = async (message) => {
...
   const responding_msg = message.reply_to_message ? message.reply_to_message : message;
+  // First handle if the message is a URL
+  const urls = responding_msg.text ? getUrls(responding_msg.text) : [];
+  if (urls.size > 0) {
+    const bot_message = await bot.sendMessage(message.chat.id, "Received URL!", {
+      reply_to_message_id: responding_msg.message_id,
+    });
+    try {
+      const [_, result] = await Promise.all([
+        bot.editMessageText("Searching using privided URL...", {
```
Comment (soruly) [inline server.js:216]:
```suggestion
        bot.editMessageText("Searching using provided URL...", {
```

**Comment by gabeklavans** [comment]
I have rebased since the last changes and made the requested changes. It was much easier to add URL searching with the new submitSearch URL-based function.

**Comment by soruly** [comment]
Please squash formatting commits and drop the commit f5be6d0312b0cbb559c2fb83400a193db20bc361 during rebase

Better to to create a separate commit solely for fixing formats / typos

**Comment by gabeklavans** [comment]
Thank you for bearing with me... I believe I have cleaned up the commit history considerably, even removing my old non-useful changes. I did a local test and it seems to still all work as intended

**Comment by gabeklavans** [comment]
Just noticed in the commits I somehow committed some of the merge conflict text... I hope you'll excuse my ignorance with properly doing a rebase 😓 but they are all not present in the latest commit, regardless.

**Comment by soruly** [comment]
Thanks, the code change looks a lot tidy now.
I'll complete the merge this weekend. Thank you <3

**Pull request merged**